worker/transfer.go

```diff
@@ -125,46 +128,107 @@ func parallelUploadSlab(ctx context.Context, sp storeProvider, shards [][]byte,
 		span.End()
 	}
 
-	// spawn workers and send initial requests
-	hostIndex := 0
+	// helper to track used hosts.
+	hostUsed := make([]bool, len(contracts))
+	nextHost := func() int {
+		for i := range hostUsed {
+			if !hostUsed[i] {
+				hostUsed[i] = true
+				return i
+			}
+		}
+		return -1
+	}
+
+	// helper to determine whether a request is finished.
+	isReqFinished := func(r req) bool {
+		select {
+		case <-r.finishedCtx.Done():
+			return true
+		default:
+			return false
+		}
+	}
+
+	// helper function for launching worker given an existing request. Returns
+	// true if a worker was launched for the request or if the request was
+	// already finished.
 	inflight := 0
-	for i := range shards {
-		go worker(req{contracts[hostIndex], i})
-		hostIndex++
+	launchWorker := func(r req) bool {
```

Comment:
I think launchWorker should receive a `req` with the `hostIndex` set and only check whether it's `-1`. That way we don't rebuild the `req` which is prone to mistakes and we avoid launching a worker with a host index of -1 and counting on the launcher to call `nextHost`.

Reply:
Rebuilding the request in there was actually done on purpose. Because we need to rebuild it anyway. So doing it in that spot seemed to make sense.

That's because we want to reset some fields and reuse others. As you already pointed out in a different comment things get tricky when you call `nextHost` outside of `launchWorker`.